docs/design/simple-rolling-update.md

```diff
@@ -0,0 +1,46 @@
+## Simple rolling update
+This is a lightweight design document for simple rolling update in ```kubectl```
+
+### Lightweight rollout
+Assume that we have a current replication controller named ```foo``` and it is running image ```image:v1```
+
+```kubectl rolling-update rc foo [foo-v2] --image=myimage:v2```
+
+Under the hood, this will do the following:
+   * Create a new replication controller named foo-next, identical to foo, except image is updated to the new image.
+   * Do the standard down/up one-by-one until foo has replicas zero and foo-next has replicas N
+
+If the user doesn't specify a name for the 'next' controller, then the 'next' controller is renamed to
+the name of the original controller as follows:
+   * delete rc foo
+   * create rc foo that is identical to foo-next
+   * delete foo-next
+
+Obviously there is a race here, where if you kill the client between delete foo, and creating the new version of 'foo' you might be surprised about what is there, but I think that's ok.
+See Recovery below
+
+If the user does specify a name for the 'next' controller, then the 'next' controller is retained with its existing name,
+and the old 'foo' controller is deleted.
+
+
+### Aborting a rollout
+Abort is assumed to want to reverse a rollout in progress.  If the rollout has completed, then rollback is just a special case of rollout.
+
+```kubectl rolling-update rc foo [foo-v2] abort```
+
+Under the hood this will:
+   * look for a replica controller named ```foo-next```, or the user-specified name (```foo-v2``` in the example above)
+   * Given this controller it will then increment/decrement until ```foo``` is back to the initial size, and ```foo-next``` is size zero.
+   * When ```foo-next``` is size zero, it is deleted.
+
+### Recovery
+To facilitate recovery in the case of a crash of the updating process itself, we add the following annotations to each replicaController:
+   * ```desired-replicas``` The desired number of replicas for this controller (either N or zero)
+   * ```update-partner``` A pointer to the resource that is the other half of this update (syntax ```<namespace>:<id>```)
+
+Recovery is achieved by issuing the 'resume' rolling update command:
```

Comment:
Agree - anything else is non-intuitive.


> On Apr 10, 2015, at 11:48 PM, Brian Grant <notifications@github.com> wrote:
> 
> In docs/design/simple-rolling-update.md:
> 
> > +### Aborting a rollout
> > +Abort is assumed to want to reverse a rollout in progress.  If the rollout has completed, then rollback is just a special case of rollout.
> > +
> > +```kubectl rolling-update rc foo [foo-v2] abort```
> > +
> > +Under the hood this will:
> > +   * look for a replica controller named ```foo-next```, or the user-specified name (```foo-v2``` in the example above)
> > +   * Given this controller it will then increment/decrement until ```foo``` is back to the initial size, and ```foo-next``` is size zero.
> > +   * When ```foo-next``` is size zero, it is deleted.
> > +
> > +### Recovery
> > +To facilitate recovery in the case of a crash of the updating process itself, we add the following annotations to each replicaController:
> > +   * ```desired-replicas``` The desired number of replicas for this controller (either N or zero)
> > +   * ```update-partner``` A pointer to the resource that is the other half of this update (syntax ```<namespace>:<id>```)
> > +
> > +Recovery is achieved by issuing the 'resume' rolling update command:
> Why can't they just execute exactly the same command?
> 
> —
> Reply to this email directly or view it on GitHub.
>